### ISSUE TITLE
pre-1.0 stability guarantees

### ISSUE THREAD

**Issue by clux** [kube-rs/kube-rs#508]
based on the [client library support level doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/csi-new-client-library-procedure.md#client-support-level) from sig api-machinery we are currently at `beta` and have the current `stable` requirements:

- [ ] support level documented per-platform
- [x] Library documentation - yes, via `#[deny(missing_docs)]`: https://github.com/clux/kube-rs/blob/38caa6beee3c7a12fc7bd46a5357e1258196cfca/kube/src/lib.rs#L74
- [ ] Deprecation policy (backwards compatibility guarantees documented)
   * How fast may the interface change?
- [x] Release process well documented - yes, [release.toml](https://github.com/clux/kube-rs/blob/master/release.toml) describes procedure, linked to in https://github.com/clux/kube-rs/commit/38caa6beee3c7a12fc7bd46a5357e1258196cfca
- [x] Versioning procedure well documented - one liner in about pre-1.0 semver strategy https://github.com/clux/kube-rs/commit/38caa6beee3c7a12fc7bd46a5357e1258196cfca
- [x] N documented users of the library - we could link to many: Stackable (all their operators) + Microsoft (krustlet, krator), [logdna agent](https://github.com/logdna/logdna-agent-v2), [rlink-rs](https://github.com/rlink-rs/rlink-rs), [kubeapps proxy](https://github.com/kubeapps/kubeapps/blob/master/cmd/pinniped-proxy/Cargo.toml), perhaps also company concepts like [h2oai](https://github.com/h2oai/h2o-kubernetes) + [ephemeron](https://github.com/qualified/ephemeron), and perhaps cli tools like [kdash](https://github.com/kdash-rs/kdash) - took a small subset in https://github.com/clux/kube-rs/commit/d8b42c79ce1f27942330d77c6dab2e1db20c51df

Some key questions and observations:
### how do we define how fast interfaces change?
obviously at post 1.0 we can deprecate fully, and parallel introduce features. but pre-1.0, things do change.
is there value in quantifying this? i know people will kind of look at these documents and use it as empty justifications to companies that's basically "hey, look, dd!". i joke, but like, how much value is there in saying "at most 2 breaking changes per release" other than it encouraging smaller releases (and we release pretty frequently in general). Maybe there are better ways to describe this?

### deprecation policies
pre 1.0 i feel we can do a best effort to not break things, and __where possible__ have deprecated functions available for at least 6 months. of course this won't always be possible yet; we don't have full feature parity with go, and the better design choices are not always clear. in those cases, we will denote changes as `BREAKING` in `CHANGELOG.md` and users can do changes at their own will through version pinning with cargo.

#### Per platform support
Well, we can document that we support Linux, and that everything else might require feedback for subtle issues. Windows / Mac are to some degree relying on strength-in-numbers atm because CI for these is in general not free.
What about Arm? We don't build for that here, but big community of tinkerers that try to build lightweight stuff on k3s.

#### definition of stability
an interesting consequence worth noting here:
because of the requirements in the linked document, we can have the `kubernetes: stable` badge even before our 1.0 release because it doesn't require us to never break things, it just requires us documenting our procedures/policies for breaking changes. Honestly, I think that's a pretty nice approach, even if it might look a bit strange to people seeing a stable badge on a pre-1.0 crate.

as a side-note following CloudNativeRustDay: people already consider us __mature__ from many shoutouts and this on twitter: https://twitter.com/JamesLaverack/status/1389326493180903427

**Comment by clux** [comment]
also potentially worth mentioning is `version compat with kubernetes`, which to a large degree is solved by `k8s-openapi` version selection, but beyond that, we mostly operate on a "what's available at the 3 big cloud providers" basis.

client-go side-steps the "how fast" question, by introducing a [compatibility matrix](https://github.com/kubernetes/client-go#compatibility-matrix), we could go down that approach, but it feels a bit unnecessary because there's technically not a lot limiting us from being used on older kubernetes versions. it's only when you do api operations on older objects.

**Comment by kazk** [comment]
> Windows / Mac are to some degree relying on strength-in-numbers atm because CI for these is in general not free.

GitHub Actions supports them for free for open source projects. We can use `strategy.matrix` to run our tests in all supported OS in parallel. For example, I've been using it to publish binaries on release. https://github.com/qualified/lsp-ws-proxy/blob/b63f30f4200ab0e6bbd273d7599a64159a4a78fc/.github/workflows/upload-binaries.yml#L12

We should be able to use `k3d` or `kind`.

**Comment by clux** [comment]
A couple of proposals/discussion points for a classes of stability guarantees:

1. When we define traits, or re-export types that rely on specific kubernetes apis (e.g. `CustomResourceExt` via #545, admission module types), we keep the old versions available until they are dropped from major cloud providers (EKS/AKS/GKE). #545 provides one potential example of this type of versioned changes. We can introduce new versions as the main versions when the are available in major cloud providers.

2. When we define __generic implementations__ that rely on core concepts of the kubernetes apis (like the watch api (kube-runtime), or the meta::v1 (affecting Api and associated traits)), then we have to make a major version change to take advantage of it.

Not sure what would happen if kubernetes dares to break meta::v1, or the watch api, but if they are going to get away with it, it sure as hell needs to be worth it. So far they've only dared to add optional metadata types, and new enum types to watchevent (under feature flags and query params), so highly doubt they'll do something drastic at this point, but hey, they might want to clean up some of the messes they have.

**Comment by clux** [comment]
Closing this in favour of #726 which is a lot more elaborate and professional.
